kernel-modules/probe/collector_probe.c

```diff
@@ -0,0 +1,340 @@
+
+  settings = get_bpf_settings();
+  if (settings == NULL) {
+    return 0;
+  }
+
+  if (!settings->capture_enabled) {
+    // capture_enabled is usually false when we have loaded the probe
+    // but not started capturing from a userspace perspective.
+    return 0;
+  }
+
+  if (id == LOOKUP_SYSCALL_ID) {
+    // this is to support sys_enter and sys_exit probes for older (RHEL 7)
+    // platforms. Just get the id from the context for this scenario.
+    id = bpf_syscall_get_nr(ctx);
+    stack_ctx.id = id;
+  }
+
+  sc_evt = get_syscall_info(id);
+  if (sc_evt == NULL) {
+    return 0;
+  }
+
+  evt_type = sc_evt->enter_event_type;
+  drop_flags = sc_evt->flags;
+
+  // To satisfy some verifier requiremnts in later parts of the falco plumbing/fillers,
+  // it is necessary to copy the context onto the stack.
+  memcpy(stack_ctx.args, _READ(ctx->args), sizeof(unsigned long) * NUM_SYS_ENTER_ARGS);
+
+  // stashing the args will copy it into a BPF map for later
+  // processing. This is a required step for the enter probe,
+  // and these args are subsequently pulled out of the map and
+  // written to the ring buffer.
+  //
+  // The args pointer must exist for the lifetime of this event, so we
+  // must not use _READ(stack_ctx.args) here.
+  if (stash_args(stack_ctx.args)) {
+    return 0;
+  }
+
+  // the fillers contain syscall specific processing logic, so we simply
+  // call into those and let the rest of falco deal with the event.
+  //
+  // It also handles the stack context problem, so we can pass both
+  // pointers through without issue.
+  call_filler(ctx, &stack_ctx, evt_type, settings, drop_flags);
+  return 0;
+}
+
+/**
+ * @brief Generic sys_exit_* program for any system call. It is responsible for
+ *        Verifying userspace settings and early processing of the syscall event.
+ *
+ *        The function will exit 0 (zero) regardless of outcome.
+ *
+ * @param id the syscall id
+ * @param ctx the eBPF context as provided by the kernel
+ *
+ * @return 0 (regardless of outcomes)
+ */
+static __always_inline int exit_probe(long id, struct sys_exit_args* ctx) {
+  const struct syscall_evt_pair* sc_evt = NULL;
+  struct sysdig_bpf_settings* settings = NULL;
+  enum ppm_event_type evt_type = PPME_GENERIC_X;
+  int drop_flags = UF_ALWAYS_DROP;
+
+  settings = get_bpf_settings();
+  if (settings == NULL) {
+    return 0;
+  }
+
+  if (!settings->capture_enabled) {
+    // capture_enabled is usually false when we have loaded the probe
+    // but not started capturing from a userspace perspective.
+    return 0;
+  }
+
+  if (id == LOOKUP_SYSCALL_ID) {
+    // this is to support sys_enter and sys_exit probes for legacy (RHEL 7)
+    // platforms. Just get the id from the context for this scenario.
+    id = bpf_syscall_get_nr(ctx);
+  }
+
+  sc_evt = get_syscall_info(id);
+  if (sc_evt == NULL) {
+    return 0;
+  }
+
+  evt_type = sc_evt->exit_event_type;
+  drop_flags = sc_evt->flags;
```

Comment:
I've just realised that this doesn't quite work now, with the possibility of this being part of a sys_enter/sys_exit probe. I'll need to check the flags in case this even is unused.